README.md
# Sidekick

Realtime site monitoring helper. Sidekick provides a few things:

  1. Integrate a pub/sub server into your server.
  1. Insight to incomming requests and the ability to react to them in code.
  1. Multiplexing http requests.  This is especially useful for testing live requests against a dev environment.
     
Please look into the examples directory for usage.

## Components in Sidekick

### Server

The sidekick server is a class that offers both the server and the middleware (connect) for your base application.

```
  var sidekick = require('./sidekick');
  sidekick.listen(7777);
  originalApp.use(sidekick.connect());
  originalApp.listen(7778);
```

At this point, you have your app running on port 7778 and a sidekick port running on 7777.  If you create a client
to access http://localhost:7777/data, you will see an incoming stream of json lines representing the requests coming 
in from the original app. 

Note: Accessing it in the browser is not a good way to view the data, you should use some kind of streaming programatic
client


### Client

The client is a way to tap into the sidekick server and subscribe to requests.

```
  var client = sidekick.Client(host, port);
  client.on('sidekick.requests', function (data) {
    /*
     * data == {
     *   headers: <hash>,
     *   path: <string>,
     *   method: <integer>
     * }
     *
     # ... do something ... 
     */
  });
```



CHANGELOG.md
### 0.0.7

  * adding subscriberCount(poolName) to server
  * adding metadata to proxy (response times)

### 0.0.6

  * adding nssocket instead of custom tcp server

### 0.0.5

  * fixing large payload bug

### 0.0.4
  
  * fixing resubscribing problem

### 0.0.2
  
  * moving over to a tcp protocol
  * setting reconnect in client
  * adding tests

### 0.0.1

  * adding hasSubscribers to sd server
 
### 0.0.1-pre8

  * adding abort to client
  * adding retry to client
  
### 0.0.1-pre7

  * pubsub
  
### 0.0.1-pre6

  * support for arbitrary data packets

### 0.0.1-pre4
  
  * support for POST/PUT requests
 
### 0.0.1-pre3

  * initial release



tests/index.js
require('mochiscript');

var connect  = require('connect');
var sidekick = require('../lib/sidekick');
var http     = require('http');


var SD_PORT    = 7777;
var ORIG_PORT  = 7778;
var CLONE_PORT = 7779;

var server = new sidekick.Server();
server.listen(SD_PORT);

// Create original app and attach sidekick server to it
var app = connect.createServer();
var count = 0;
app.use(server.connect());
app.use(connect.query());
app.use(function (req, res, next) { 
  server.publish(req.url);
  server.publish('count', ++count); 
  res.end(req.url);
});
app.listen(ORIG_PORT);

var delayN = 0;

var helper = {
  request: function (path) {
    http.get({ host: 'localhost', port: ORIG_PORT, path: path }, function (res) {});
  },

  server: function () {
    return server;
  },

  client: function () {
    var client = new sidekick.Client('localhost', SD_PORT);
    return client;
  },

  delay: function (cb, n) {
    n = n ? n : delayN += 200;
    setTimeout(cb, n);
  }
}


require('./stop')(helper);
require('./no-connect')(helper);
require('./subscriber-count')(helper);


examples/server.js
/*
 * A full on example of how to use sidekick as a proxy multiplexer
 * 1) create sd server
 * 2) create app
 * 3) create clone port
 * 4) create sd client
 * 5) create a proxy 
 * 6) add proxy to client
 */
var connect  = require('connect');
var sidekick = require('../lib/sidekick');
var http     = require('http');

var SD_PORT    = 7777;
var ORIG_PORT  = 7778;
var CLONE_PORT = 7779;

// Create sidekick server
var sd = new sidekick.Server();
sd.listen(SD_PORT);

// Create original app and attach sidekick server to it
var origApp = connect.createServer();
origApp.use(sd.connect());
origApp.use(function (req, res, next) { 
  setTimeout(function () { 
    console.log('Orig Server: Get request!');
    res.end("hello");
  }, 200);
});
origApp.listen(ORIG_PORT);

// Create clone app
var cloneApp = connect.createServer();
cloneApp.use(function (req, res, next) { 
  setTimeout(function () { 
    console.log('Clone app: Get request!');
    res.end("hello");
  }, 200);
});
cloneApp.listen(CLONE_PORT);

// Create sidekick client to listen to sidekick server
var client = new sidekick.Client('localhost', SD_PORT);
client.on('response', function (data) { 
  console.log("Sidekick Client: " + JSON.stringify(data));
});

// Create proxy
var proxy = client.proxy('localhost', CLONE_PORT);
proxy.on('response', function (code, body, headers) {
  console.log("Proxy: got response.");
});

setInterval(function () { 
  console.log("External Request: Make request to original server");
  http.get({ host: 'localhost', port: ORIG_PORT, path: '/' }, function (res) {});
}, 1000);


examples/counter.js
/*
 * A full on example of how to use sidekick as a proxy multiplexer
 * 1) create sd server
 * 2) create app
 * 5) create a proxy to multiplex the request
 */
var connect  = require('connect');
var sidekick = require('../lib/sidekick');
var http     = require('http');

var SD_PORT    = 7777;
var ORIG_PORT  = 7778;

// Create sidekick server
var sd = new sidekick.Server();
sd.listen(SD_PORT);

// Create original app and attach sidekick server to it
var origApp = connect.createServer();
var count   = 0;
origApp.use(sd.connect());
origApp.use(function (req, res, next) { 
  setTimeout(function () { 
    console.log('Orig Server: Get request!');
    sd.publish("request-count", ++count);

    res.end("hello");
  }, 200);
});
origApp.listen(ORIG_PORT);


// Create sidekick client to listen to sidekick server
var client = new sidekick.Client('localhost', SD_PORT);
client.subscribe('request-count', function (data) { 
  console.log("Number of Requests: " + data);
});


setInterval(function () { 
  console.log("External Request: Make request to original server");
  http.get({ host: 'localhost', port: ORIG_PORT, path: '/' }, function (res) {});
}, 1000);


examples/multiplexer.js
/*
 * A full on example of how to use sidekick as a proxy multiplexer
 * 1) create sd server
 * 2) create app
 * 3) create clone app
 * 4) create sd client
 * 5) create a proxy to multiplex the request
 */
var connect  = require('connect');
var sidekick = require('../lib/sidekick');
var http     = require('http');

var SD_PORT    = 7777;
var ORIG_PORT  = 7778;
var CLONE_PORT = 7779;

// Create sidekick server
var sd = new sidekick.Server();
sd.listen(SD_PORT);

// Create original app and attach sidekick server to it
var origApp = connect.createServer();
var count = 0;
origApp.use(sd.connect());
origApp.use(function (req, res, next) { 
  setTimeout(function () { 
    console.log('Orig Server: Get request!');
    sd.publish('count', ++count); 
    res.end("hello");
  }, 200);
});
origApp.listen(ORIG_PORT);

// Create clone app to run in parallel with original app
var cloneApp = connect.createServer();
cloneApp.use(function (req, res, next) { 
  setTimeout(function () { 
    console.log('Clone app: Get request!');
    res.end("hello");
  }, 200);
});
cloneApp.listen(CLONE_PORT);

// Create sidekick client to listen to sidekick server
var client = new sidekick.Client('localhost', SD_PORT);
client.subscribe('sidekick.requests', function (data) { 
  console.log("Sidekick Client: " + JSON.stringify(data));
});

client.subscribe('count', function (data) {
  console.log("Sidekick Client Count: " + data);
});

// Create proxy to send requests to cloned app
var proxy = client.proxy('localhost', CLONE_PORT);
proxy.on('response', function (code, body, headers, meta) {
  console.log("Proxy: got response in " +  meta.responseTime + 'ms');
  console.log("--------------------------");
});

setInterval(function () { 
  console.log("External Request: Make request to original server");
  http.get({ host: 'localhost', port: ORIG_PORT, path: '/' }, function (res) {});
}, 1000);



examples/test-large-payload.js
/*
 * A full on example of how to use sidekick as a proxy multiplexer
 * 1) create sd server
 * 2) create app
 * 3) create clone app
 * 4) create sd client
 * 5) create a proxy to multiplex the request
 */
var connect  = require('connect');
var sidekick = require('../lib/sidekick');
var http     = require('http');

var assert     = require('assert');
var longString = "/";
for (var i=0; i<80000; i++) longString += "a";

var SD_PORT    = 7777;
var ORIG_PORT  = 7778;
var CLONE_PORT = 7779;

// Create sidekick server
var sd = new sidekick.Server();
sd.listen(SD_PORT);

// Create original app and attach sidekick server to it
var origApp = connect.createServer();
var count = 0;
origApp.use(sd.connect());
origApp.use(function (req, res, next) { 
  setTimeout(function () { 
    console.log('Orig Server: Get request!');
    sd.publish('count', ++count); 
    res.end("hello");
  }, 200);
});
origApp.listen(ORIG_PORT);

// Create clone app to run in parallel with original app
var cloneApp = connect.createServer();
cloneApp.use(function (req, res, next) { 
  setTimeout(function () { 
    console.log('Clone app: Get request!');
    res.end("hello");
  }, 200);
});
cloneApp.listen(CLONE_PORT);

// Create sidekick client to listen to sidekick server
var client = new sidekick.Client('localhost', SD_PORT);
client.subscribe('sidekick.requests', function (data) { 
  console.log("Sidekick Client: asserting...");
  assert.equal(data.path, longString);
});

client.subscribe('count', function (data) {
  console.log("Sidekick Client Count: " + data);
});

// Create proxy to send requests to cloned app
var proxy = client.proxy('localhost', CLONE_PORT);
proxy.on('response', function (code, body, headers) {
  console.log("Proxy: got response.");
  console.log("--------------------------");
});

setInterval(function () { 
  console.log("External Request: Make request to original server");
  http.get({ host: 'localhost', port: ORIG_PORT, path: longString }, function (res) {});
}, 1000);

